glide-core/src/client/value_conversion.rs

```diff
@@ -891,7 +892,115 @@ pub(crate) fn convert_to_expected_type(
+                2# "creation_timestamp" => 1728348101740745
+                3# "key_type" => "HASH"
+                4# "key_prefixes" =>
+                   1) "blog:post:"
+                5# "fields" =>
+                   1) 1# "identifier" => "category"
+                      2# "field_name" => "category"
+                      3# "type" => "TAG"
+                      4# "option" => ""
+                   2) 1# "identifier" => "vec"
+                      2# "field_name" => "VEC"
+                      3# "type" => "TAVECTORG"
+                      4# "option" => ""
+                      5# "vector_params" =>
+                         1# "algorithm" => "HNSW"
+                         2# "data_type" => "FLOAT32"
+                         3# "dimension" => 2
+                ...
+
+            Map keys (odd array elements) are simple strings, not bulk strings.
+            */
+            Value::Array(_) => {
+                let Value::Map(mut map) = convert_to_expected_type(value, Some(ExpectedReturnType::Map {
+                    key_type: &None,
+                    value_type: &None,
+                }))? else { unreachable!() };
+                for pair in map.iter_mut() {
+                    if pair.0 == Value::SimpleString("fields".into()) {
+                        let Value::Array(mut fields) = pair.1.clone() else {
+                            return Err((
+                                ErrorKind::TypeError,
+                                "Response couldn't be converted for FT.INFO",
+                                format!("(`fields` was {:?})", get_value_type(&pair.1.clone())),
+                            )
+                                .into());
+                        };
+
+                        for field in fields.iter_mut() {
+                            let Value::Map(mut field_params) = convert_to_expected_type(field.clone(), Some(ExpectedReturnType::Map {
+                                key_type: &None,
+                                value_type: &None,
+                            })).unwrap() else { unreachable!() };
+
+                            for pair in field_params.iter_mut() {
+                                if pair.0 == Value::SimpleString("vector_params".into()) {
+                                    *pair = (pair.0.clone(), convert_to_expected_type(pair.1.clone(), Some(ExpectedReturnType::Map {
+                                        key_type: &None,
+                                        value_type: &None,
+                                    }))?);
+                                    break;
+                                }
+                            }
+
+                            *field = Value::Map(field_params);
+                        }
+
+                        *pair = (pair.0.clone(), Value::Array(fields));
+                        break;
+                    }
+                }
+                Ok(Value::Map(map))
```

Comment:
As requested, I dug into this a bit more to see if I could reduce the number of clones here. Here's what I've come up with: 
```suggestion
let Some(fields_pair) = map.iter_mut().find(|(key, _)| {
                    *key == Value::SimpleString("fields".into())
                }) else { return Ok(Value::Map(map)) };
                let (fields_key, fields_value) = std::mem::replace(fields_pair, (Value::Nil, Value::Nil));
                let Value::Array(fields) = fields_value else {
                    return Err((
                        ErrorKind::TypeError,
                        "Response couldn't be converted for FT.INFO",
                        format!("(`fields` was {:?})", get_value_type(&fields_value)),
                    ).into());
                };

                let fields = fields.into_iter().map(|field| {
                    let Value::Map(mut field_params) = convert_to_expected_type(field, Some(ExpectedReturnType::Map {
                        key_type: &None,
                        value_type: &None,
                    }))? else { unreachable!() };

                    let Some(vector_params_pair) = field_params.iter_mut().find(|(key, _)| {
                        *key == Value::SimpleString("vector_params".into())
                    }) else { return Ok(Value::Map(field_params)) };
                    let (vector_params_key, vector_params_value) = std::mem::replace(vector_params_pair, (Value::Nil, Value::Nil));

                    let _ = std::mem::replace(vector_params_pair, (vector_params_key, convert_to_expected_type(vector_params_value, Some(ExpectedReturnType::Map {
                        key_type: &None,
                        value_type: &None,
                    }))?));

                    Ok(Value::Map(field_params))
                }).collect::<RedisResult<Vec<Value>>>()?;

                let _ = std::mem::replace(fields_pair, (fields_key, Value::Array(fields)));
                Ok(Value::Map(map))
```

I gathered some stats to show that the number of allocations is reduced from 36 to 23. You might be wondering if that `collect` call allocates memory. I checked, and it does not. The number of allocations is still 23, even if you use a mem replace instead.

Here's what heaptrack produces when I run this code in a small sample app:

**Before**
![image](https://github.com/user-attachments/assets/d821dd4e-14f6-4ec9-898a-565c6b0597fc)

**After**
![image](https://github.com/user-attachments/assets/4a039970-5daa-4b13-9bd4-c5f1559732bb)

Here is the sample app code (I expose glide_core client types and functions in public API just for testing purposes):
```rust
use redis::Value;
use glide_core::client::{ExpectedReturnType, convert_to_expected_type};

fn main() {
    let value = Value::Array(vec![
        Value::Array(vec![Value::SimpleString("fields".to_string()), Value::Array(vec![
            Value::Array(vec![
                Value::SimpleString("vector_params".to_string()),
                Value::Array(vec![Value::SimpleString("hi".to_string()), Value::Int(42)])
            ])
        ])])
    ]);
    convert_to_expected_type(value, Some(ExpectedReturnType::FTInfoReturnType)).unwrap();
}
```

Reply:
Updated, thank you for the review!